File: add-two-numbers/yd.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return helper(l1, l2, 0);
    }
    
    private ListNode helper(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null && carry == 0) {
            return null;
        }
        carry += l1 == null ? 0 : l1.val;
        carry += l2 == null ? 0 : l2.val;
        ListNode ret = new ListNode(carry % 10);
        ret.next = helper(l1 == null ? null : l1.next, l2 == null ? null : l2.next, carry / 10);
        return ret;
    }
}

File: add-two-numbers/yd.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        return self.helper(l1, l2, 0)
    
    def helper(self, l1, l2, c):
        if l1 is None and l2 is None and c == 0:
            return None
        c += 0 if l1 is None else l1.val
        c += 0 if l2 is None else l2.val
        node = ListNode(c % 10)
        node.next = self.helper(None if l1 is None else l1.next, None if l2 is None else l2.next, c / 10)
        return node
